File: 094_binary_tree_inorder_traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        inOrder(root, result);
        return result;
    }

    // iterative solution
//     public List<Integer> inorderTraversal(TreeNode root) {
//         List<Integer> res = new ArrayList();
//         Stack<TreeNode> stack = new Stack();
//         while(!stack.isEmpty() || root != null){
//             while(root != null){
//                 stack.push(root);
//                 root = root.left;
//             }
//             root = stack.pop();
//             res.add(root.val);
//             root = root.right;
//         }

//         return res;
//     }

    private void inOrder(TreeNode node, List<Integer> result) {
        if (node == null) return;

        inOrder(node.left, result);
        result.add(node.val);
        inOrder(node.right, result);
    }

}

File: python/094_binary_tree_inorder_traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        res = []

        stack = []
        while root:
            stack.insert(0, root)
            root = root.left

        while len(stack) > 0:
            node = stack.pop(0)
            res.append(node.val)

            right = node.right
            while right:
                stack.insert(0, right)
                right = right.left
        return res
